tracker/src/main/java/org/matomo/sdk/dispatcher/DefaultDispatcher.java

```diff
@@ -139,6 +151,44 @@ public boolean forceDispatch() {
         return true;
     }
 
+    @Override
+    public void forceDispatchBlocking() {
+        Thread dispatchThread = null;
+
+        synchronized (mThreadControl) {
+            // ensure the dispatch loop always exits after a single loop
+            mForcedBlocking = true;
+
+            if (mRunning) {
+                dispatchThread = mDispatchThread;
+                mRetryCounter = 0;
+                mSleepToken.release();
+            }
+        }
+
+        if (dispatchThread != null) {
+            try {
+                dispatchThread.join();
+            } catch (InterruptedException e) {
+                Timber.tag(TAG).i("Interrupted while waiting for dispatch thread to complete");
+            }
+        }
+
+        if (!mEventCache.isEmpty()) {
+            synchronized (mThreadControl) {
+                mRunning = true;
+                mSleepToken.release();
+            }
+
+            mLoop.run();
```

Comment:
Thats a good point. In my testing the dispatcher was set offline before making the blocking call so no events were sent to network.  But there is no guarantee that offline will be called first.

In the worst case it's running the loop a second time but the events are already on disk and no work will be done.  If there is no dispatcher loop running something needs to force the code through the submit loop to ensure they are all written out.  I would have gated the call by checking to see if the event cache has events left in it but the current `isEmpty` implementation includes events on disk and would report more work to do if the dispatcher was offline.

I tried to limit the scope of the api changes and this seemed like an inexpensive way to avoid changing `EventCache` code and still ensure the events are written out.

I think this one needs some tweaking to avoid the network issue...

Reply:
Removed this in [b2bf532](https://github.com/matomo-org/matomo-sdk-android/pull/252/commits/b2bf532a0232453cbc3d8c9bd02b5bdfd46edddd)